<a href="https://colab.research.google.com/github/Ritesh-panda/datascience-learnings/blob/main/Bagging.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [37]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.ensemble import BaggingClassifier, RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.datasets import make_classification

In [19]:
x,y=make_classification(n_samples=10000,n_features=10,n_informative=3)

In [20]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2, random_state=42)

In [21]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
x_train_scaled=scaler.fit_transform(x_train)
x_test_scaled=scaler.transform(x_test)

In [22]:
model_dt=DecisionTreeClassifier()
model_dt.fit(x_train_scaled,y_train)

DecisionTreeClassifier()

In [23]:
y_pred_dt=model_dt.predict(x_test_scaled)
y_pred_dt

array([0, 1, 0, ..., 0, 1, 0])

In [24]:
print("decosion tree accuracy",accuracy_score(y_test,y_pred_dt)*100)

decosion tree accuracy 89.75


In [32]:
bag=BaggingClassifier(
    estimator=DecisionTreeClassifier(), # The base model to clone (Decision Tree)
    n_estimators=10,                  # Number of trees in the ensemble
    max_samples=0.5,                 # Train each tree on 50% of the data
    bootstrap=True,                   # Samples are drawn with replacement
    random_state=42                  # Ensures reproducible results
)
bag

BaggingClassifier(estimator=DecisionTreeClassifier(), max_samples=0.5,
                  random_state=42)

In [33]:
# Fitting the Bagging model
bag.fit(x_train_scaled, y_train)

BaggingClassifier(estimator=DecisionTreeClassifier(), max_samples=0.5,
                  random_state=42)

In [34]:
y_pred_bag=bag.predict(x_test_scaled)
y_pred_bag

array([0, 1, 0, ..., 0, 1, 0])

In [35]:
print("Bagging Classifier accuracy",accuracy_score(y_test,y_pred_bag)*100)

Bagging Classifier accuracy 92.10000000000001


**Random forest**

In [39]:
model_RF=RandomForestClassifier()
model_RF.fit(x_train_scaled,y_train)
y_pred_RF=model_RF.predict(x_test_scaled)
y_pred_RF
print("RandomForest",accuracy_score(y_test,y_pred_RF)*100)

RandomForest 92.80000000000001


In [42]:
bag_RF=BaggingClassifier(
    estimator=RandomForestClassifier(), # The base model to clone (RF)
    n_estimators=10,                  # Number of trees in the ensemble
    max_samples=0.5,                 # Train each tree on 50% of the data
    bootstrap=True,                   # Samples are drawn with replacement
    random_state=42                  # Ensures reproducible results
)
bag_RF

BaggingClassifier(estimator=RandomForestClassifier(), max_samples=0.5,
                  random_state=42)

In [44]:
bag_RF.fit(x_train_scaled, y_train)
y_pred_bag_RF=bag_RF.predict(x_test_scaled)
y_pred_bag_RF

array([0, 1, 0, ..., 0, 1, 0])

In [45]:
print("Bagging Classifier WITH RANDOM FOREST accuracy",accuracy_score(y_test,y_pred_bag_RF)*100)

Bagging Classifier WITH RANDOM FOREST accuracy 92.4


**Bagging using SVM**

In [49]:
bag_svm=BaggingClassifier(
    estimator=SVC(), # The base model to clone (SVM)
    n_estimators=10,                  # Number of trees in the ensemble
    max_samples=0.25,                 # Train each tree on 50% of the data
    bootstrap=True,                   # Samples are drawn with replacement
    random_state=42                  # Ensures reproducible results
)
bag_svm.fit(x_train_scaled,y_train)

BaggingClassifier(estimator=SVC(), max_samples=0.25, random_state=42)

In [50]:

y_pred_bag_svm=bag_svm.predict(x_test_scaled)
y_pred_bag_svm

array([0, 1, 0, ..., 0, 1, 0])

In [51]:
print("Bagging Classifier WITH svm accuracy",accuracy_score(y_test,y_pred_bag_svm)*100)

Bagging Classifier WITH svm accuracy 90.14999999999999


**Pasting**
in bagging you have the bootstaping but in pasting you dont have the bootstrapping

In [52]:
# Implementing Pasting (bootstrap=False)
pasting_clf = BaggingClassifier(
    estimator=DecisionTreeClassifier(),
    n_estimators=10,
    max_samples=0.5,
    bootstrap=False, # This makes it Pasting instead of Bagging
    random_state=42
)

# Training the model
pasting_clf.fit(x_train_scaled, y_train)

# Making predictions
y_pred_pasting = pasting_clf.predict(x_test_scaled)

# Evaluating the model
print("Pasting Classifier accuracy:", accuracy_score(y_test, y_pred_pasting) * 100)

Pasting Classifier accuracy: 92.45


### **Key Takeaways from Bagging & Pasting**

1. **Ensemble Power**: Both Bagging and Pasting improved the performance compared to a single **Decision Tree (~89.75%)**. By training multiple models on different subsets of data, we reduced the variance and achieved higher accuracy (**~92%+**).

2. **Bagging (Bootstrap Aggregating)**:
   * **Mechanism**: Uses sampling **with replacement** (`bootstrap=True`).
   * **Diversity**: Because the same instance can appear multiple times in a single subset, it introduces more diversity in the training of individual base learners.
   * **Result**: Achieved **92.1%** accuracy with Decision Trees.

3. **Pasting**:
   * **Mechanism**: Uses sampling **without replacement** (`bootstrap=False`).
   * **Subset size**: Each model sees a unique subset of the data.
   * **Result**: Achieved **92.45%** accuracy in this specific run. Often, Pasting works well when you have a large dataset.

4. **Flexibility of Base Estimators**: We demonstrated that the `BaggingClassifier` is versatile; it can wrap around **Decision Trees**, **Random Forests**, or even **SVMs** to create an ensemble.

5. **Random Forest vs. Bagging DT**: The `RandomForestClassifier` (**92.8%**) is essentially a specialized version of Bagging that also selects a random subset of features at each split, which often provides the best balance of bias and variance.